workers/unity/Packages/com.improbable.gdk.deploymentlauncher/DeploymentTasks.cs

```diff
@@ -0,0 +1,267 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Improbable.Gdk.Tools;
+using UnityEditor;
+using UnityEngine;
+
+namespace Improbable.Gdk.DeploymentManager
+{
+    internal static class DeploymentTasks
+    {
+        private static readonly string DotNetProjectPath =
+            Path.GetFullPath(Path.Combine(Tools.Common.GetPackagePath("com.improbable.gdk.deploymentlauncher"),
+                ".DeploymentLauncher/DeploymentLauncher.csproj"));
+
+        private static TimeSpan DeploymentLauncherTimeout = new TimeSpan(0, 0, 25, 0);
+
+
+        public class WrappedTask<TResult> : IDisposable
+        {
+            public Task<TResult> Task;
+            public CancellationTokenSource CancelSource;
+
+            public void Dispose()
+            {
+                Task?.Dispose();
+                CancelSource?.Dispose();
+            }
+        }
+
+        public static WrappedTask<bool> TriggerUploadAssemblyAsync(string[] arguments, Action onSuccess,
+            Action onFailure)
+        {
+            var source = new CancellationTokenSource();
+            var token = source.Token;
+
+            var projectRootPath = Path.Combine(Application.dataPath, "../../../");
+
+            var task = Task.Run(async () =>
+            {
+                if (!Tools.Common.CheckDependencies())
+                {
+                    return false;
+                }
+
+                var processResult = await RedirectedProcess.Command(Tools.Common.SpatialBinary).WithArgs(arguments)
+                    .RedirectOutputOptions(OutputRedirectBehaviour.RedirectStdOut |
+                        OutputRedirectBehaviour.RedirectStdErr | OutputRedirectBehaviour.ProcessSpatialOutput)
+                    .InDirectory(projectRootPath)
+                    .RunAsync(token);
+
+                if (processResult.ExitCode == 0)
+                {
+                    onSuccess();
+                    return true;
+                }
+
+                onFailure();
+                return false;
+            });
+
+            return new WrappedTask<bool>
+            {
+                Task = task,
+                CancelSource = source
+            };
+        }
+
+        public static WrappedTask<bool> TriggerLaunchDeploymentAsync(List<string> arguments, Action onSuccess)
```

Comment:
shouldn't the arguments be inside these methods? otherwise we can't be sure that `TriggerLaunchDeploymentAsync` is actually triggering the launch of a deployment. likewise for the other methods

Reply:
The context for these arguments is built inside the editor window and passed in here. These can be made class methods on the editor window, but I'd rather start abstracting the UI from the business logic

Reply:
but only parts of the argument, right? e.g. deployment name and project name. just pass those in directly instead and then construct the arguments inside of this method